### ISSUE TITLE
Implementation of Chameleon?

### ISSUE THREAD

**Issue by jimbethancourt** [haifengl/smile#139]
Hi!
Thanks for the amazing work you've done.  Are there any plans to add the Chameleon hierarchical clustering algorithm by Karypis, Han, and Kumar to the list of algorithms in the collection?  I've noticed you have several Chameleon datasets under source control and you have the K Nearest Neighbor data structure already developed.

Thanks,
Jim

**Comment by haifengl** [comment]
Currently no plan to support it. Would you like to implement it? I am more than happy to help. Thanks!

**Comment by jimbethancourt** [comment]
Possibly -- I've found a Java implementation at http://www.cnblogs.com/zhangchaoyang/articles/2182752.html but the author doesn't list a license, so I'm not sure if it would be appropriate to base an implementation on it.

Thoughts?

Thanks,
Jim

**Comment by haifengl** [comment]
I wouldn't take the risk of license. But it seems not very complicated and we have many built in functions. It should be easy to implement it.

**Comment by deric** [comment]
In order to have the original Chamelenon as described in the paper, you'd need a [hMETIS](http://glaros.dtc.umn.edu/gkhome/metis/hmetis/download) implementation that is available only as binary. Also the license states that it is for non-commercial usage only.

**Comment by SomeUserName1** [comment]
Karger-Stein's Algorithm in it's parallel version (see [section 7](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=2ahUKEwj9lYCohd7hAhXIalAKHYsiDE0QFjAAegQIAhAC&url=http%3A%2F%2Fwww.columbia.edu%2F~cs2035%2Fcourses%2Fieor6614.S09%2FContraction.pdf&usg=AOvVaw3BhDdzgaA-tZr4s1Fyi64_)) provides partitioning by finding min-cuts which is the strategy mentioned in the chameleon paper by [Kapis et al.](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=3&ved=2ahUKEwiP08bRhd7hAhWO-aQKHUNnDLkQFjACegQIAxAC&url=https%3A%2F%2Fwww-users.cs.umn.edu%2F~han%2Fdmclass%2Fchameleon.pdf&usg=AOvVaw0OiJ83nBeZ2ILzUZwYmUiC) for both partitioning and merging (the metric for merging includes finding min-cut bisectors efficiently).
Thus Chameleon would be kNN, followed by min-cut partition, followed by merging with a fancy metric (Rel.I = |EC c_i, c_j| / |E c_i| + |E c_j|; Rel.C = the same but rather summing over the weights of the edges for each edge cut sub-term and normalizing wrt. internal closeness)

In other words, when finished with HDBScan, I'd give it a shot

**Comment by jimbethancourt** [comment]
Hi - I noticed this was closed.  I'm guessing there are no plans to implement?

Thanks,
Jim

**Comment by haifengl** [comment]
No plan

**Comment by jimbethancourt** [comment]
Understood - thanks for considering it.

Best regards,
Jim
